### ISSUE TITLE
NoReturn sometimes does not get narrowed when referenced inside a closure

### ISSUE THREAD

**Issue by rslinckx** [microsoft/pyright#3551]
This is with pyright 1.1.252

```
from flask import Flask, abort

app = Flask(__name__)


def get_something_or_none() -> str | None:
    return None


@app.route("/")
def index():
    x = get_something_or_none()
    if not x:
        abort(404)

    def foo() -> str:
        return x.upper() # <- upper is not a known member of "None"

    return x.upper() # <- upper is not a known member of "None"
```

while this version type-checks properly (I just removed the inline foo() definition):

```
from flask import Flask, abort

app = Flask(__name__)


def get_something_or_none() -> str | None:
    return None


@app.route("/")
def index():
    x = get_something_or_none()
    if not x:
        abort(404)

    return x.upper()
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

In general, variables captured by an inner function or lambda to not take on a narrowed type from the outer scope. For example, mypy never does this because it's not type safe in general. However, pyright adds some special-case logic to allow the narrowed type to be inherited by the inner function or lambda in specific cases where it can be proven that the captured variable is not reassigned after the function or lambda declaration. Proving this is quite complex, which explains why mypy hasn't yet implemented this feature. The logic involved in this code path contained a bug which explains what you were seeing.

**Comment by rslinckx** [comment]
Thanks ! That explains also why I stumbled across it while investigating the previous bug, and couldn't figure out why the behaviour changed by just having the closure. The most intriguing part was that the closure made the outer scope complain about the type as well, even though the closure itself never touched the variable, but i guess this is quite complex to determine statically

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.253, which I just published. It will also be included in the next release of pylance.
